scripts/build_test.py

```diff
@@ -14,29 +14,270 @@
+        """Test _compare_file_count to raise exception when there is a
+        mismatched file count between 2 dirs.
+        """
+        with self.assertRaises(ValueError) as incorrectFileCount:
+            build._compare_file_count(
+                build.EXTENSIONS_DIRNAMES_TO_DIRPATHS['dev_dir'],
+                build.TEMPLATES_CORE_DIRNAMES_TO_DIRPATHS['dev_dir'])
+        source_dir_file_count = build.get_file_count(
+            build.EXTENSIONS_DIRNAMES_TO_DIRPATHS['dev_dir'])
+        target_dir_file_count = build.get_file_count(
+            build.TEMPLATES_CORE_DIRNAMES_TO_DIRPATHS['dev_dir'])
+        self.assertTrue(
+            ('%s files in source dir != %s files in target dir.') % (
+                source_dir_file_count, target_dir_file_count) in
+            incorrectFileCount.exception)
+
+    def test_match_filename_with_hashes(self):
+        """Test _match_filename_with_hashes to raise exception:
+            1) When there is an empty hash dict.
+            2) When a filename is expected to contain hash but does not.
+            3) When there is a hash in filename that cannot be found in
+                hash dict.
+        """
+        # Final filepath example: base.240933e7564bd72a4dde42ee23260c5f.html.
+        file_hashes = dict()
+        base_filename = 'base.html'
+        with self.assertRaises(ValueError) as emptyHashDict:
+            build._match_filename_with_hashes(base_filename, file_hashes)
+        self.assertTrue('Hash dict is empty' in emptyHashDict.exception)
+
+        file_hashes = {base_filename: random.getrandbits(128)}
+        bad_filepath = 'README'
+        with self.assertRaises(ValueError) as filepathWithoutExtension:
+            build._match_filename_with_hashes(bad_filepath, file_hashes)
+        # Generate a random hash dict for base.html.
+        self.assertTrue(
+            'Filepath has less than 2 partitions after splitting'
+            in filepathWithoutExtension.exception)
+
+        with self.assertRaises(ValueError) as noHashInFilename:
+            build._match_filename_with_hashes(base_filename, file_hashes)
+        # Generate a random hash dict for base.html.
+        self.assertTrue(
+            '%s is expected to contain MD5 hash' % base_filename
+            in noHashInFilename.exception)
+
+        hashed_base_filename = build._insert_hash(
+            base_filename, random.getrandbits(128))
+        with self.assertRaises(KeyError) as incorrectHashInFilename:
+            build._match_filename_with_hashes(hashed_base_filename, file_hashes)
+        self.assertTrue(
+            'Hash from file named %s does not match hash dict values'
+            % hashed_base_filename in incorrectHashInFilename.exception)
+
+    def test_process_html(self):
+        """Test process_html to remove whitespaces and hash filepaths."""
+        base_source_path = os.path.join(
+            build.TEMPLATES_CORE_DIRNAMES_TO_DIRPATHS['dev_dir'], 'pages',
```

Comment:
Again, for this, use a test build dir in core/tests/build. You can define some prod dir in there too and .gitignore it, and maybe nuke it at the start of each test run. Put some sample files there that are representative of our codebase in general; the test should make sure they are correctly processed.

That way, if people change files in the main codebase, these tests will not break.

Reply:
Done!